sstables/trie_writer.hh

```diff
@@ -0,0 +1,1183 @@
+// The complicated part is the size calculations needed to group parents and children together.
+//
+// For the most part we follow Cassandra's way of doing that.
+// See also:
+// https://github.com/apache/cassandra/blob/f9e2f1b219c0b570eb59d529da42722178608572/src/java/org/apache/cassandra/io/tries/IncrementalTrieWriterPageAware.java#L32
+// https://github.com/apache/cassandra/blob/f9e2f1b219c0b570eb59d529da42722178608572/src/java/org/apache/cassandra/io/sstable/format/bti/BtiFormat.md?plain=1#L347
+// The main difference is that our writer operates in units of character chains,
+// rather than on single characters.
+// With a node per character, the performance of a writer based on single characters
+// can get catastrophically bad for long keys.
+
+#pragma once
+
+#include <seastar/util/log.hh>
+#include <map>
+#include <set>
+#include "utils/assert.hh"
+#include "utils/small_vector.hh"
+
+extern seastar::logger trie_logger;
+
+namespace trie {
+
+// Enables code which is useful for debugging during development,
+// but too expensive to be compiled into release builds (even if dynamically disabled).
+constexpr bool developer_build = false;
+// Adds expensive extra checks against use-after-free on pointers obtained from the bump_allocator.
+// Must be a macro so that it can affect whether some struct fields are defined or not.
+#define TRIE_SANITIZE_BUMP_ALLOCATOR developer_build
+// Many asserts and logs are only useful during development,
+// where the cost of logging doesn't matter at all.
+// And during development it might be useful to have A LOT of them,
+// and make them expensive.
+//
+// But in production, if the code is hot enough is frequent enough, we might care even about the small
+// cost of dynamically checking whether a logger is enabled, which discourages adding more trace logs,
+// which is sad?
+//
+// Maybe the right thing is to do is to have a special most verbose log level disabled at compile time by default,
+// and only enabled by developers who are actively working on the relevant feature?
+// This way we free to add as many logs as we please, without worrying at all about the performance cost.
+template <typename... Args>
+void expensive_log(seastar::logger::format_info_t<Args...> fmt, Args&&... args) {
+    if constexpr (developer_build) {
+        trie_logger.trace(std::move(fmt), std::forward<Args>(args)...);
+    }
+}
+inline void expensive_assert(bool expr, std::source_location srcloc = std::source_location::current()) {
+    if constexpr (developer_build) {
+        if (!expr) {
+            __assert_fail("", srcloc.file_name(), srcloc.line(), srcloc.function_name());
+        }
+    }
+}
+
+// Yet another way to represent a buffer view.
+// We aleady have bytes_view, and perhaps I should use it instead.
+//
+// But std::span<const std::byte> is, in some sense, the standard type for this purpose.
+// std::as_bytes() exists, after all.
+using const_bytes = std::span<const std::byte>;
```

Comment:
Using `std::span<const std::byte>` here requires the bytes to be linearized. Could we use `managed_bytes_view` instead?